File: sol/solution/0001-0100/0030. Substring with Concatenation of All Words/Solution.dart
class Solution {
  List<int> findSubstring(String s, List<String> words) {
    if (s.isEmpty || words.isEmpty) return [];
    int wordLen = words[0].length;
    int totalLen = wordLen * words.length;
    Map<String, int> wordCount = {};
    for (var w in words) {
      wordCount[w] = (wordCount[w] ?? 0) + 1;
    }
    List<int> res = [];
    for (int i = 0; i < wordLen; i++) {
      int left = i, count = 0;
      Map<String, int> seen = {};
      for (int j = i; j + wordLen <= s.length; j += wordLen) {
        String word = s.substring(j, j + wordLen);
        if (wordCount.containsKey(word)) {
          seen[word] = (seen[word] ?? 0) + 1;
          count++;
          while (seen[word]! > wordCount[word]!) {
            String leftWord = s.substring(left, left + wordLen);
            seen[leftWord] = seen[leftWord]! - 1;
            left += wordLen;
            count--;
          }
          if (count == words.length) res.add(left);
        } else {
          seen.clear();
          count = 0;
          left = j + wordLen;
        }
      }
    }
    return res;
  }
}

File: sol/solution/0001-0100/0030. Substring with Concatenation of All Words/Solution.kt
class Solution {
    fun findSubstring(s: String, words: Array<String>): List<Int> {
        if (words.isEmpty()) return emptyList()
        val wordLen = words[0].length
        val totalLen = wordLen * words.size
        val wordCount = words.groupingBy { it }.eachCount()
        val res = mutableListOf<Int>()
        for (i in 0 until wordLen) {
            var left = i
            var count = 0
            val seen = mutableMapOf<String, Int>()
            var j = i
            while (j + wordLen <= s.length) {
                val word = s.substring(j, j + wordLen)
                if (wordCount.containsKey(word)) {
                    seen[word] = seen.getOrDefault(word, 0) + 1
                    count++
                    while (seen[word]!! > wordCount[word]!!) {
                        val leftWord = s.substring(left, left + wordLen)
                        seen[leftWord] = seen[leftWord]!! - 1
                        left += wordLen
                        count--
                    }
                    if (count == words.size) res.add(left)
                } else {
                    seen.clear()
                    count = 0
                    left = j + wordLen
                }
                j += wordLen
            }
        }
        return res
    }
}